### PR TITLE
[WIP] Draft for makefile commands + docs update; #297 #307 #308.

### PR DESCRIPTION
**Pull request by TimDettmers**
This introduces a draft for download data and preprocessing files with make commands. It would be easy for users to use this, and it would be easy for users to extend this. Currently, it is a mess and it takes ages to understand how to run a simple model. I know Jack and even for me it took ages. This is not good. Let us also discuss different options how to make this better.

Currently the make command for SQuAD is broken, because the processing is not done correctly; see #308.

### CODE REVIEW

**Comment by TimDettmers** [comment]
See also the `make list` command to list all models. This is more immediate so that users can use the same procedure, make file calls, to (1) find out what they can run, (2) do data download and preprocessing, (3) run simple models. I think this would be a good way to go.

File: Makefile [hunk 1]
```diff
@@ -24,7 +24,54 @@ overfit:
...
-SNLI:
-	$(PYTEST) tests -v -m SNLI
+
+# ===========
+# SNLI
+# ===========
+
+snli:
+	./data/SNLI/download.sh
+	$(PYTHON) jack/io/SNLI2jtr.py
```
Comment (dirkweissenborn) [inline Makefile]: We don't need the conversion anymore. There are loaders for snli now.

File: Makefile [hunk 2]
```diff
@@ -24,7 +24,54 @@ overfit:
...
+snli-esim:
+	 $(PYTHON) bin/jack-train.py with \
+		 train=data/SNLI/snli_1.0/snli_1.0_train_jtr_v1.json \
+		dev=data/SNLI/snli_1.0/snli_1.0_dev_jtr_v1.json \
+		test=data/SNLI/snli_1.0/snli_1.0_test_jtr_v1.json \
```
Comment (dirkweissenborn) [inline Makefile]:
```
loader=snli
train=data/SNLI/snli_1.0/snli_1.0_trai.json
dev=data/SNLI/snli_1.0/snli_1.0_dev.json
test=data/SNLI/snli_1.0/snli_1.0_test.json
```

File: Makefile [hunk 3]
```diff
@@ -24,7 +24,54 @@ overfit:
...
+		model=esim_snli_reader
+
+
+# ===========
+# SQuAD
+# ===========
+#
+squad:
+	./data/SQuAD/download.sh
```
Comment (dirkweissenborn) [inline Makefile]: cd into SQuAD, see README

File: Makefile [hunk 4]
```diff
@@ -24,7 +24,54 @@ overfit:
...
+
+glove:
+	./data/GloVe/download.sh
+
+squad-fastqa:
+	 $(PYTHON) bin/jack-train.py with \
+		 train=data/SQuAD/
+		dev=data/SNLI/snli_1.0/snli_1.0_dev_jtr_v1.json \
```
Comment (dirkweissenborn) [inline Makefile]:
```
bin/jack-train.py with config='./conf/fastqa.yaml'
```

File: Makefile [hunk 5]
```diff
@@ -24,7 +24,54 @@ overfit:
...
+
+# ===========
+# SNLI
+# ===========
+
+snli:
+	./data/SNLI/download.sh
+	$(PYTHON) jack/io/SNLI2jtr.py
+snli-esim:
+	 $(PYTHON) bin/jack-train.py with \
```
Comment (dirkweissenborn) [inline Makefile]: we should never use the standard jack.yaml config, but inherit from it and create new configs for each model or at least task. E.g. you could create a `conf/natural_language_inference.yaml` which by default trains a model on snli. See the `conf/fastqa.yaml` for instance. in the command line we only need to overwrite the model type (e.g., esim, bilstm, ...)

File: Makefile (hunk 3 above)
Comment (dirkweissenborn) [reply Makefile]: but we should change that st we do not have to cd

**Comment by dirkweissenborn** [comment]
@TimDettmers I like this PR after addressing the minor issues I mentioned and fixing some of the problems you had (e.g., fix download scripts). Is there anything else that is blocking this PR? I think we should get that in soon.

**Comment by dirkweissenborn** [comment]
I updated readme and the download scripts, should be better now

**Comment by TimDettmers** [comment]
If we want to use the `Makefile` as "documentation" is may make sense to include both types of invocation, that is via config and via command line parameters. What do you think?

I will address the changes later today or tomorrow.

**Comment by dirkweissenborn** [comment]
Whatever you think is better. Personally, I think the using task specific configs with minor overwrites on the command line are nice. That means that the task config gives you a reasonable training setup (hyperparameters) such that you only need to define the model, model_dir etc on the command line.

**Comment by TimDettmers** [comment]
Hmm, that is true. I like the config + overwrite better. However, we would need to work on something that makes clear to the user what parameters are available and what is their syntax/name. Some models have different parameter options than others. How do we generalize here?

Maybe its better to just list all of them for a start? Just `--help parameters` or something, and we also have `--help models` and `--help datasets` and when you type `--help` it just returns a list of these commands, what do you think?

**Comment by undwie** [comment]
Is this good to merge? Maybe we can add a "config introspection" issue for 0.2?

**Comment by dirkweissenborn** [comment]
@undwie this is not yet ready, see my comments (although partially outdated) above.

**Comment by TimDettmers** [comment]
I fixed the SNLI commands now, but the SQuAD commands run on errors even after rebase. I think some embedding config value is missing, but this is not clear from the error and not very user-friendly. I also tried the command that you proposed and it also does not because my memory blew up (I think this is GloVe, is it?). Can you fix this @dirkweissenborn?

@undwie we should fix these issues before a release, otherwise people cannot really use jack. If someone that worked on jack (me) cannot use it after 10 minutes of trying to figure out how to run something without a config then this is a pretty bad design. Also, the first release is crucial for long-term success.

**Comment by undwie** [comment]
Note that 0.1 wasn't supposed to be the "let's get it out there release" but a pre-step to that. I am mentioning this because we run the risk of constantly moving goal posts. 

But yes, what you say generally true! My point was about ideas like "listing all configuration items automatically" etc. which are great ideas, but I think good examples/readme's can deal with this for the time being. I managed to run `FastQA` relatively easily by just looking at the configuration files and readme. Maybe that's because I am okay with using config files (if I am explicitly said to use these). 

In any case, we should improve usability dramatically, but maybe not all in 0.1? In fact, my wish for 0.2 would be to mostly focus on usability improvements!

File: Makefile [hunk 6]
```diff
@@ -24,7 +24,56 @@ overfit:
...
+#
+squad:
+	./data/SQuAD/download.sh
+
+glove:
+	./data/GloVe/download.sh
+
+squad-fastqa:
+	#$(PYTHON) bin/jack-train.py with config='./conf/fastqa.yaml'
+	$(PYTHON) bin/jack-train.py with \
```
Comment (dirkweissenborn) [inline Makefile]: embeddings need to be defined here. FastQA only works when embeddings are provided. Just use the config statement instead. The README shows how to use the commandline instead

**Pull request closed without merging**